### DISCUSSION TITLE
fixture teardown does not seems to execute

### DISCUSSION THREAD

**Discussion by tr-gis** [pytest-dev/pytest#9996] [General] [answered]
My tests directory structure is as follows:
```

tests
|---- conftest.py
|---- helper.py
|---- object1
      |----conftest.py
      |----test_object1.py
|---- object2
      |----conftest.py
      |----test_object2.py
|---- object3
      |----conftest.py
      |----test_object3.py
.
.
.
.

```

contents of conftest.py on the top level is as follows:

```python
import pytest
from .helper import *


@pytest.fixture(scope='session')
def object_manager():
    def _dispatcher(object_name, data, method ): 
           
        response = send_request(object_name, data, method)   <--- creates object
        print('Before yield')
        yield response
        print('After yield')
        response = send_request(object_name, data, method)    <-- deletes object
            
    return _dispatcher
```
I use this fixture in test_object1.py within object1 directory. The contents of test_object1.py is as follows:

```python
import pytest

def test_setup(object_manager):
    response = next(object_manager(object_name=POD, data=values_for_new_object()))
    assert response['success'] 
```
Then I run pytest with options to show setup for fixtures and capture set to no using the following command:
`pytest --setup-show -s` and got the following output:

```
================================================================================= test session starts ==================================================================================
platform linux -- Python 3.8.10, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /home/abs/Desktop/user/project/tests
plugins: anyio-3.3.4
collected 1 item                                                                                                                                                                       

object1/test_object1.py 
SETUP    S object_manager
        object1/test_object1.py::test_setup (fixtures used: object_manager)Before yield
.
TEARDOWN S object_manager

================================================================================== 1 passed in 0.10s ===================================================================================
```

From this I can see that my fixture entered it's SETUP phase, where it created the object like it is supposed to, it even prints the 'Before yield' statement I put in there for debugging. Next the asserts in my test_setup function pass, I get the green dot too. 

However when the fixture enters TEARDOWN phase it does not seem to delete the object. There is nothing wrong with send_request() function , I checked. The statement "After yield" should be printed when the fixture enters TEARDOWN phase, I do not see it in the output. I feel like all the code below yield are not executed. Could you please tell me what went wrong here ?

**Comment by nicoddemus** [comment, marked as answer]
Hi @tr-gis,

> Could you please tell me what went wrong here ?

Your fixture is returning a function, which you call on the test directly. From the fixture's point of view, there's no teardown phase, because you return an object from it (the `object_manager` function), and that's it. The fixture can't know that your function yields a value, and that you expect that the function should continue and execute the code after the yield as *teardown*. The function object has been returned from the fixture, and pytest has no control over it anymore.

The `yield` mechanism (setup until `yield`, teardown after) only works if the `yield` statement is in the body of the fixture function; in your example, your fixture body contains just a function (with a `yield` inside it), but the fixture cannot know that: from its point of view, it returned an object, and that's it.



From your example, it doesn't seem like you need a fixture at all, and could just use a context-manager instead:

```python
# Implemented in some module importable by your app, say `app.testing`.
@contextmanager
def object_manager(object_name, data, method):        
    response = send_request(object_name, data, method)
    print("Before yield")
    try:
        yield response
    finally:
        print("After yield")
        response = send_request(object_name, data, method)
        
```

```python
from app.testing import object_manager

def test_setup():
    with object_manager(object_name=POD, data=values_for_new_object()) as response:    
        assert response['success'] 
```

Hope this helps.

**Reply by tr-gis** [reply]
Hi @nicoddemus . Thank you for your response. This worked beautifully however I want the objects created to last for the entire session so that test functions written for other objects (Note: I have updated my directory structure on the very top of this post to reflect this) can refer to them as well .

With the code snippet that you provided :

```
from app.testing import object_manager

def test_setup():
    with object_manager(object_name=POD, data=values_for_new_object()) as response:    
        assert response['success'] 

```

The object would be destroyed as soon as I exit test_setup() function. This will create problems since the plan is to structure code like below :

```
# let this file be test_object1.py
from app.testing import object_manager

def test_setup():
    with object_manager(object_name=POD, data=values_for_new_object()) as response:    
        assert response['success'] 
       
    ##### say object created is called POD1, we assign it's id to some global variable say POD1_ID

def test_update_object_params():
    ##### update param of object with id POD1_ID
    this fails since the context manager cleaned up the object

```
and say I am referring to this from another tests file:

```
# let this file be object2/ test_object2.py
from app.testing import object_manager

def test_setup():
    with object_manager(object_name=STARSHIP, data=POD1_ID) as response:    <-- this would fail since POD1 was deleted by context manager and does not exist anymore
        assert response['success'] 

```

**Reply by nicoddemus** [reply]
I see. Well I would strongly advise *against* having tests depend on data created by other tests, this creates an interdependence between them which is bad for maintainability and understanding the suite, tests are supposed to be independent from one another.  


Having said that, it is possible, but you need to create a class that will hold your IDs, and the fixture will provide an instance of that class. For example:


```python

class ObjectManager:

    def open_object(self, name: str, data: str) -> Response:
        """Return a response for the given object"""

    def set_object(self, id: str, response: Response) -> None:
        """Stores a response with an id."""

    def get_object(self, id: str) -> Response:
        """Return the response obtained from a previous ``set_object`` call."""

    def close(self):
        """Cleans up all open objects."""


@pytest.fixture(scope="session")
def manager():
    manager = ObjectManager()
    yield manager
    manager.close()


def test_setup(manager):
    response = manager.open_object(name=POD, data=values_for_new_object())
    assert response['success'] 

    manager.set_object(POD1, response)
           

def test_update_object_params(manager):
    response = manager.get_object(POD1)
    response.update_param("foo")
```

But as I said, **I don't recommend this, making tests depend on one another is not a good practice**.

It is better to just have each test indepently create and operate on each "object". If what you want to share between the tests is some long operations done in an object, you can easily instead move that code to a fixture or function.

**Reply by tr-gis** [reply]
Hm. Well I guess I will have to abandon this path if this is not a good practice. Thank you for your time @nicoddemus . :smile:
